indexed_gzip/zran.c

```diff
@@ -2384,3 +2384,453 @@ eof:         return ZRAN_READ_EOF;
+         * imported for all elements except the first one.
+         */
+        if (point != new_list) {
+
+            /*
+             * Allocate space for checkpoint data. These pointers in each point
+             * should be cleaned up in case of any failures.
+             */
+            point->data = calloc(1, index->window_size);
+
+            if (point->data == NULL) goto memory_error;
+
+            /*
+             * Read checkpoint data, and check for errors. End of file can be
+             * reached just after the last element, so it's not an error for
+             * the last element.
+             */
+            f_ret = fread(point->data, index->window_size, 1, fd);
+
+            if (feof(fd) && point < list_end - 1) goto eof;
+            if (ferror(fd))                       goto read_error;
+            if (f_ret != 1)                       goto read_error;
+
+            /*
+             * TODO: If there are still more data after importing is done, it
+             * is silently ignored. It might be handled by other means.
+             */
+        }
+
+        #ifdef ZRAN_VERBOSE
+        zran_log("zran_import_index: (cmp: %lu, uncmp: %lu, bits: %u, ",
+                 point->cmp_offset,
+                 point->uncmp_offset,
+                 point->bits);
+
+        /* First point supposed to not have a checkpoint window. */
+        if (point->data != NULL) {
+
+            /* Print first and last three bytes of the checkpoint window. */
+            zran_log("window: [%02x %02x %02x...%02x %02x %02x])\n",
+                     point->data[0],
+                     point->data[1],
+                     point->data[2],
+                     point->data[index->window_size - 3],
+                     point->data[index->window_size - 2],
+                     point->data[index->window_size - 1]);
+        } else {
+            zran_log("window: NULL)\n");
+        }
+        #endif
+
+        /* Done with this point. Proceed to the next one. */
+        point++;
+
+    }
+
+    /* There are no errors, it's safe to overwrite existing index data now. */
+
+    /* If a new uncompressed_size is read, update current index. */
+    if (index->uncompressed_size == 0 && uncompressed_size != 0) {
+        index->uncompressed_size = uncompressed_size;
+    }
+
+    /* Overwrite spacing. */
+    if (index->spacing != spacing) {
+        index->spacing = spacing;
+    }
+
+    /* Overwrite window size. */
+    if (index->window_size != window_size) {
+        index->window_size = window_size;
+    }
+
+    /*
+     * Now, we will release current checkpoint list of the index, and then
+     * point to the new list.
+     */
+
+    /*
+     * Initialize point to the second element of the list, and list_end to the
+     * end of the list. We initialize to the second element, because first
+     * element does not keep checkpoint data.
+     */
+    point    = index->list + 1;
```

Comment:
How innocent this line looks, isn't it? :)

Below while silently fails to run because of this mistake.